### PR TITLE
Detection of local extrema from morphology

### PR DESCRIPTION
**Pull request by ThomasWalter**
## Description
Detection of local maxima and minima in an image. So far, skimage allows to detect local maxima with the function `peak_local_max `, but in some circumstances, this function is not optimal. For instance, it is impossible to retrieve all local maxima/minima of an image, irrespective of any distance or threshold consideration. Also, the definition of a local maximum as a set of pixels larger than any other pixel within a distance, does not suit all applications. Here, we use the morphological definition of a local maximum as a set of pixels strictly greater/lower than all surrounding pixels. We also define the h-minima and h-maxima, i.e. minima with a depth of at least h and maxima with a height of at least h. This is based on the morphological reconstruction in the morphology module implemented by Lee Kamentsky. All functions can be applied to all image types. The strategy was already mentioned in the help of the reconstruction algorithm, but I made this more formal and applicable to all data types. 

## Checklist
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
[If this is a bug-fix or enhancement, it closes issue # ]
[If this is a new feature, it implements the following paper: ]
I do not remember the first paper mentioning this, but it is based on a standard textbook: 

Soille, P. (2003). Morphological Image Analysis: Principles and Applications
(2nd ed.). Springer-Verlag New York, Inc.

### CODE REVIEW

File: skimage/morphology/tests/test_extrema.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+            data = data.astype(dtype)
+            out = extrema.h_minima(data, 40)
+
+            error = diff(expected_result, out)
+            assert error < eps
+            assert out.dtype == expected_result.dtype
+
+    def test_extrema_float(self):
+        "specific tests for float type"
+        print 'test_extrema_float'
```
Comment (sciunto) [inline skimage/morphology/tests/test_extrema.py]: Please, remove this line.

File: skimage/morphology/extrema.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+This module provides functions to find local maxima and minima of an image.
+Here, local maxima (minima) are defined as connected sets of pixels with equal
+grey level which is strictly greater (smaller) than the grey level of all
+pixels in direct neighborhood of the connected set. In addition, the module
+provides the closely related functions h-maxima and h-minima.
+
+Soille, P. (2003). Morphological Image Analysis: Principles and Applications
+(2nd ed.). Springer-Verlag New York, Inc.
+
+Original author: Thomas Walter
```
Comment (sciunto) [inline skimage/morphology/extrema.py]: Instead, we prefer that you add your name in CONTRIBUTORS.txt :)

File: skimage/morphology/extrema.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+import numpy as np
+
+
+def _add_constant_clip(img, const_value):
+    """Adds a constant to the image and handles overflow issues
+    for integer typed images.
+    """
+    array_info = np.iinfo(img.dtype)
+    max_dtype = array_info.max
+    min_dtype = array_info.min
```
Comment (sciunto) [inline skimage/morphology/extrema.py]: in skimage.util.dtype we have dtype_limits that does this job.

File: skimage/morphology/extrema.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+    decreasing by more than h with respect to the maximum's value) and no
+    path for which the minimal value is greater.
+
+    Parameters
+    ----------
+    img : ndarray
+        The input image for which the maxima are to be calculated.
+    h : unsigned integer
+        The minimal height of all extracted maxima.
+    selem : ndarray
```
Comment (sciunto) [inline skimage/morphology/extrema.py]: , optional

File: skimage/morphology/extrema.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    increasing by more than h with respect to the minimum's value) and no
+    path for which the maximal value is smaller.
+
+    Parameters
+    ----------
+    img : ndarray
+        The input image for which the minima are to be calculated.
+    h : unsigned integer
+        The minimal depth of all extracted minima.
+    selem : ndarray
```
Comment (sciunto) [inline skimage/morphology/extrema.py]: , optional

File: skimage/morphology/extrema.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+
+    For integer typed images, this corresponds to the h-maxima with h=1.
+    For float typed images, h is determined as the smallest difference
+    between grey levels.
+
+    Parameters
+    ----------
+    img : ndarray
+        The input image for which the maxima are to be calculated.
+    selem : ndarray
```
Comment (sciunto) [inline skimage/morphology/extrema.py]: , optional

File: skimage/morphology/extrema.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+
+    For integer typed images, this corresponds to the h-minima with h=1.
+    For float typed images, h is determined as the smallest difference
+    between grey levels.
+
+    Parameters
+    ----------
+    img : ndarray
+        The input image for which the minima are to be calculated.
+    selem : ndarray
```
Comment (sciunto) [inline skimage/morphology/extrema.py]: same here.

File: skimage/morphology/extrema.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+pixels in direct neighborhood of the connected set. In addition, the module
+provides the closely related functions h-maxima and h-minima.
+
+Soille, P. (2003). Morphological Image Analysis: Principles and Applications
+(2nd ed.). Springer-Verlag New York, Inc.
+
+Original author: Thomas Walter
+"""
+
+import greyreconstruct
```
Comment (sciunto) [inline skimage/morphology/extrema.py]: from skimage.morphology import greyreconstruct

File: skimage/morphology/tests/test_extrema.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+import numpy as np
+
+from skimage.morphology import extrema
+
+eps = 1e-12
+
+
+def diff(a, b):
+    a = np.asarray(a)
+    a = a.astype(np.float64)
```
Comment (sciunto) [inline skimage/morphology/tests/test_extrema.py]: can't we merge as
np.asarray(a, dtype=np.float64) ?

File: skimage/morphology/tests/test_extrema.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    b = np.asarray(b)
+    b = b.astype(np.float64)
```
Comment (sciunto) [inline skimage/morphology/tests/test_extrema.py]: same here.

**Comment by ThomasWalter** [comment]
Thanks for your comments; I will correct this tomorrow. Just one question concerning the example I would need to provide: can I add an image ? How do I do this exactly ?

**Comment by sciunto** [comment]
solution 1: use an image in our module data
solution 2: build an image in the example
solution 3; add an image in our data module (for that, we need something sufficiently generic to use it in multiple examples) and with a compatible license (If you are the authors, noting to do).

examples are in doc/examples.

**Comment by sciunto** [comment]
You need to rebase on master as well.

File: skimage/morphology/watershed.py [hunk 11]
```diff
@@ -255,8 +258,13 @@ def watershed(image, markers, connectivity=1, offset=None, mask=None,
...
                                  mask, image_strides, compactness,
-                                 output.ravel())
+                                 output.ravel(),
+                                 watershed_line)
 
     output = crop(output, pad_width, copy=True)
 
+    if watershed_line:
+        min_val = output.min()
+        output[output == min_val] = 0
```
Comment (sciunto) [inline skimage/morphology/watershed.py]: These two lines can be merged.

File: skimage/morphology/extrema.py [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+    See also
+    --------
+    skimage.morphology.extrema.h_minima
+    skimage.morphology.extrema.local_maxima
+    skimage.morphology.extrema.local_minima
+
+    References
+    ----------
+    .. [1] Soille, P., "Morphological Image Analysis: Principles and
+           Applications", 2nd edition (2003), ISBN 3540429883.
```
Comment (sciunto) [inline skimage/morphology/extrema.py]: is there a range of specific pages you can provide to guide the reader?

**Comment by jni** [comment]
@ThomasWalter it looks like your previous PR (watershedline) has been swept into this one. You will need to:

- grab the latest master branch. Not merge (which would cause our histories to diverge), but grab/check-out.
- check out a new branch
- cherry pick the right commits from this PR
- force push that branch to overwrite these changes.

Other than that, awesome contribution, thanks! I'm excited to have this in scikit-image.

File: skimage/morphology/extrema.py (hunk 8 above)
Comment (jni) [inline skimage/morphology/extrema.py]: That's not right either:

```python
from . import greyreconstruct
```

You might also need to add

```python
from __future__ import absolute_import
```
at the very top of the file.

File: skimage/morphology/extrema.py [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+        shifted_img = img - h
+        resolution = np.finfo(img.dtype).resolution
+    else:
+        shifted_img = _subtract_constant_clip(img, h)
+        resolution = 0
+
+    rec_img = greyreconstruct.reconstruction(shifted_img, img,
+                                             method='dilation', selem=selem)
+    residue_img = img - rec_img
+    result_img = np.zeros(img.shape)
```
Comment (jni) [inline skimage/morphology/extrema.py]: Specify the dtype here, rather than calling "astype" later

File: skimage/morphology/extrema.py [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+        The input image for which the maxima are to be calculated.
+    h : unsigned integer
+        The minimal height of all extracted maxima.
+    selem : ndarray
+        The neighborhood expressed as a 2-D array of 1's and 0's.
+        Defaults to a 3x3 square (8-connectivity).
+
+    Returns
+    -------
+    h-maxima : ndarray
```
Comment (jni) [inline skimage/morphology/extrema.py]: the return value should be a valid Python variable name

File: skimage/morphology/extrema.py [hunk 15]
```diff
@@ -0,0 +1,349 @@
...
+    path for which the minimal value is greater.
+
+    Parameters
+    ----------
+    img : ndarray
+        The input image for which the maxima are to be calculated.
+    h : unsigned integer
+        The minimal height of all extracted maxima.
+    selem : ndarray
+        The neighborhood expressed as a 2-D array of 1's and 0's.
```
Comment (jni) [inline skimage/morphology/extrema.py]: I think this can be n-dimensional, no?

**Comment by ThomasWalter** [comment]
Sorry for the mess with the branches. I am not sure having done the right thing though. 

- `git checkout master`
- `git fetch upstream`
- `git pull upstream master`
- `git branch -m extrema extrema_from_wrong_master`
- `git branch extrema`
- `git checkout extrema`
- `git checkout extrema_from_wrong_master skimage/morphology/extrema.py`
- `git checkout extrema_from_wrong_master skimage/morphology/tests/test_extrema.py`

I did not understand what you mean with 
> Not merge (which would cause our histories to diverge), but grab/check-out.

**Comment by sciunto** [comment]
looks correct to me. You can consider the comments and at then end squash you commit and change the commit message ;)

File: skimage/morphology/extrema.py (hunk 2 above)
Comment (ThomasWalter) [inline skimage/morphology/extrema.py]: Sorry. I did not want to get particular credits, but I had seen that in other files. I remove this of course.

File: skimage/morphology/extrema.py (hunk 8 above)
Comment (ThomasWalter) [inline skimage/morphology/extrema.py]: For me, all 3 worked fine. I do not know the reasons, but I stay with the last solution.

File: skimage/morphology/extrema.py (hunk 3 above)
Comment (ThomasWalter) [inline skimage/morphology/extrema.py]: Very handy. Thanks.

File: skimage/morphology/extrema.py (hunk 15 above)
Comment (ThomasWalter) [inline skimage/morphology/extrema.py]: Probably. `selem` is passed to `greyreconstruct` (so I do not access this myself), I therefore took the description from `greyreconstruct`. I did not see a nD test, but as far as I can tell from the code, it should work in n dimensions. I will make a 3D test to see whether this truly works and we can then change the two descriptions.

File: skimage/morphology/extrema.py (hunk 2 above)
Comment (sciunto) [inline skimage/morphology/extrema.py]: np. Usually, we do it when a code is used from a third-party.

File: skimage/morphology/extrema.py (hunk 15 above)
Comment (ThomasWalter) [inline skimage/morphology/extrema.py]: I checked this and added a test: this works in 3D. I also updated the description of greyreconstruct.py accordingly.

File: skimage/morphology/extrema.py [hunk 16]
```diff
@@ -0,0 +1,350 @@
...
+
+
+from __future__ import absolute_import
+import numpy as np
+from . import greyreconstruct
+from skimage.util import dtype_limits
+
+def _add_constant_clip(img, const_value):
+    """Adds a constant to the image and handles overflow issues
+    for integer typed images.
```
Comment (jni) [inline skimage/morphology/extrema.py]: (should be present imperative — add, not adds — and fit within 80 characters.)

File: skimage/morphology/extrema.py [hunk 17]
```diff
@@ -0,0 +1,350 @@
...
+                         "with the image data type.")
+
+    result = img + const_value
+    result[img > max_dtype-const_value] = max_dtype
+    return(result)
+
+
+def _subtract_constant_clip(img, const_value):
+    """Subtracts a constant from the image and handles overflow issues
+    for integer typed images.
```
Comment (jni) [inline skimage/morphology/extrema.py]: """Subtract constant from image while handling underflow issues."""

File: skimage/morphology/extrema.py [hunk 18]
```diff
@@ -0,0 +1,350 @@
...
+        raise ValueError("The subtracted constant is not compatible"
+                         "with the image data type.")
+
+    result = img - const_value
+    result[img < (const_value + min_dtype)] = min_dtype
+    return(result)
+
+
+def h_maxima(img, h, selem=None):
+    """Determines all maxima of the image with height >= h.
```
Comment (jni) [inline skimage/morphology/extrema.py]: Determine

File: skimage/morphology/extrema.py [hunk 19]
```diff
@@ -0,0 +1,350 @@
...
+                                             method='erosion', selem=selem)
+    residue_img = rec_img - img
+    h_minima = np.zeros(img.shape, dtype=np.uint8)
+    h_minima[residue_img >= h - 10*resolution] = 1
+    return h_minima
+
+
+def find_min_diff(img):
+    """
+    Finds the minimal difference of grey levels inside the image.
```
Comment (jni) [inline skimage/morphology/extrema.py]: """Find minimal difference of grey levels within the image.""" (one line)

File: skimage/morphology/extrema.py [hunk 20]
```diff
@@ -0,0 +1,350 @@
...
+    """
+    img_vec = np.unique(img.flatten())
+    img_vec.sort()
+    min_diff = np.min(img_vec[1:] - img_vec[:-1])
+    return min_diff
+
+
+def local_maxima(img, selem=None):
+    """Determines all local maxima of the image.
```
Comment (jni) [inline skimage/morphology/extrema.py]: Determine

File: skimage/morphology/extrema.py [hunk 21]
```diff
@@ -0,0 +1,350 @@
...
+
+Soille, P. (2003). Morphological Image Analysis: Principles and Applications
+(2nd ed.), Chapter 6. Springer-Verlag New York, Inc.
+"""
+
+
+from __future__ import absolute_import
+import numpy as np
+from . import greyreconstruct
+from skimage.util import dtype_limits
```
Comment (ThomasWalter) [inline skimage/morphology/extrema.py]: Okay, thanks a lot for the clarification!

**Comment by ThomasWalter** [comment]
Thanks again for your comments. 

By adding an example, I discovered a bug due to the floating point precision. Indeed, I am not really used to mathematical morphology applied to float images. 

The tests and the example run fine now. 

One question: morphology is mostly about assessing the order of values. The question is how do I handle the precision of float in these cases. In this particular case, I needed to find the pixels >= h in the image `residue_img` and set them to 1 in the result image `h_max`. For integer typed arrays, that's trivial: 

`h_max[residue_img >= h] = 1`

For float typed images this is less clear, as `residue_img >= h` might be wrong because of a lack in precision (this generated a bug in the code). This is particularly annoying in this algorithm, as the maximum of `residue_img` is h by construction. 

In this particular case, there are two possibilities: 

1. I can correct for the precision. But how? 

`h_corrected = h - np.finfo(img.dtype).resolution`
`h_max[residue_img >= h_corrected] = 1 # this is what I did` 

or 

`h_corrected = h - np.finfo(img.dtype).eps`
`h_max[residue_img >= h_corrected] = 1` 

2. I could just transform images to integer. The actual values do usually not matter, only their order. I did not go for this, but it might lead to code that is actually easier to understand. 

What is your feeling about this?

**Comment by stefanv** [comment]
This is a much needed contribution; thank you @ThomasWalter.  I like your epsilon / resolution approach; it strikes me as being cleaner than the hard-handed approach of converting to int.

**Comment by stefanv** [comment]
Kudos on the very neat PR; please also rebase (we fixed some Travis-CI issues earlier today).

**Comment by ThomasWalter** [comment]
Hello, 

thanks for your comments! 

So, I got some trouble with that eps ... actually, the numpy documentation is not very clear on that point. 

[https://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.finfo.html](url)

I had neglected that obviously the floating point resolution depends on the value itself (which can be retrieved by `np.spacing(x)`). But as images in skimage are supposed to contain only values between [0,1], we can use `np.finfo(dtype).eps` which is the same as `np.spacing(1.0)` as a worst case scenario. However, this does not indicate up to which decimal the calculation can be assumed to be _correct_ (and in practice it also did not work in my code). If I take  `np.finfo(type).resolution`, this seems to work, even though it is unclear to me, what this actually means (as resolution/precision should also depend on the value, as eps does). 

It remains a bit enigmatic ... but for this PR it should be fine for now.  

Just one question: what do you mean by 

> please also rebase (we fixed some Travis-CI issues earlier today).

Could you tell me what exactly this refers to? Thx !

**Comment by jni** [comment]
@ThomasWalter 

This is what it means to rebase:
http://scikit-image.org/docs/dev/contribute.html#divergence-between-upstream-master-and-your-feature-branch

Regarding the floating point accuracy, why go for floating point resolution? The algorithm instead has a resolution, the minimum difference, so, if I'm not mistaken, you could simply subtract `min_diff/2`?

And as another point that I missed earlier, the output of `np.unique` is guaranteed to be sorted, so you don't need the call to `.sort()`.

**Comment by ThomasWalter** [comment]
Thanks. I corrected the `np.unique` issue. 

`min_diff` can be actually smaller than the floating point precision (`np.array([1e-25]) - np.array([1e-26]) < np.finfo(a.dtype).eps`): if there are very small values `min_diff` can thus take values for which `f(x) + min_diff != f(x)` does not hold for all values `f(x)` in the image. 

So, I need to lower bound the algorithm's resolution to the worst precision I might find for the image values. And this is for floating point scikit images `np.spacing(1.0)` which is the same as `np.finfo(a.dtype).eps`. I could also use `np.spacing(img.max())` which would then allow me to apply the algorithm on images with values larger than 1.0. 

After that, I need to identify all pixels for which `R_f(f+h)-f == h` (for minima detection, R_f(g): reconstruction of g over f) or equivalently `R_f(f+h)-f >= h` (as by construction `R_f(f+h)-f` is upper bounded by `h`). I relax this condition to `R_f(f+h)-f >= h-eps`. I could also set this to `R_f(f+h)-f >= h-min_diff/2`.

**Comment by ThomasWalter** [comment]
Thanks also for the link, I rebased the branch.

File: skimage/morphology/extrema.py [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+
+    rec_img = greyreconstruct.reconstruction(shifted_img, img,
+                                             method='erosion', selem=selem)
+    residue_img = rec_img - img
+    h_min = np.zeros(img.shape, dtype=np.uint8)
+    h_min[residue_img >= h_corrected] = 1
+    return h_min
+
+
+def find_min_diff(img):
```
Comment (sciunto) [inline skimage/morphology/extrema.py]: Is it intended to be a public function?

File: doc/examples/segmentation/plot_extrema.py [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+We detect local maxima in a galaxy image. The image is corrupted by noise,
+generating many local maxima. To keep only those maxima with sufficient
+local contrast, we use h-maxima.
+"""
+import matplotlib.pyplot as plt
+
+from skimage.measure import label
+from skimage import data
+from skimage import color
+import numpy as np
```
Comment (sciunto) [inline doc/examples/segmentation/plot_extrema.py]: Put this import on top.

File: skimage/morphology/extrema.py (hunk 22 above)
Comment (ThomasWalter) [inline skimage/morphology/extrema.py]: Sorry, I somehow missed them. Surely private; I change the name accordingly.

File: doc/examples/segmentation/plot_extrema.py (hunk 23 above)
Comment (ThomasWalter) [inline doc/examples/segmentation/plot_extrema.py]: OK

**Comment by sciunto** [comment]
Almost! :) You just need to fix "SyntaxError: from __future__ imports must occur at the beginning of the file"

To squash, it's almost the same procedure as rebasing. You can check that on the web, it's fairly easy.

**Comment by sciunto** [comment]
Approved as well. Let's wait 0.14 to merge new features. Many thanks!

**Comment by ThomasWalter** [comment]
Hello everybody, 

I was just wondering: when do you think will we merge this? 
Because I am currently developing connected operators (for which I need local extrema). I guess that I can have a PR on this next week.

**Comment by sciunto** [comment]
There are only 1 or 2 remaining issues for the next release. So pretty soon I hope. @jni will confirm, he's the release manager. In the mean time, you can develop on your side and keep it in the freezer :)

**Comment by ThomasWalter** [comment]
OK cool. 

Thanks a lot !

**Comment by jni** [comment]
I'm in favour of merging this regardless of the state of 0.13. If it's ready, it's one more feature we get to ship! Anyone opposed, @scikit-image/core?

**Comment by sciunto** [comment]
We can go, I'm not strictly opposed.

**Comment by jni** [comment]
@ThomasWalter thanks very much for your contribution! I'm very excited to have this in skimage. =)

**Comment by soupault** [comment]
We should add those functions to `__init__.py`.

**Pull request merged**